### PR TITLE
[MRG+1] encode unicode selector exception

### PR DESCRIPTION
**Pull request by Digenis**
The exception quotes an xpath string
which may be unicode.

When an invalid xpath contains wide characters
eg ```sel.xpath(u'//x[@y="з"/]')``` (this is a Cyrilic "z", not a "3")
they end up verbatim in the exception message (which quotes the xpath).
Unicode strings should be encoded http://bugs.python.org/issue2517
Currently a UnicodeEncodeError (eg ``` UnicodeEncodeError('ascii', u'Invalid XPath: //x[@y="\u0437"/]', 23, 24, 'ordinal not in range(128)')```) hides the original exception.

~~In this PR I encode the exception message to utf8 and update the relevant test.~~
In this PR I encode the exception message to unicode_escape in python2 and update the relevant tests.

### CODE REVIEW

**Comment by Digenis** [comment]
Back to the issue.
See how I used `... if six.PY3 else ...`, some people may disagree with this.

Also, I was considering `scrapy.utils.python.unicode_to_str` but it seems it's not updated for python3 compatibility. I can try taking care of this, maybe in the same PR.

**Comment by kmike** [comment]
@Digenis IMHO `.. if six.PY3 else ...` is fine as-is. The PR looks good, but I think it is better to split `test_invalid_xpath` into two tests, one for unicode message and one for ascii-safe message. You can add `assertInvalidXpathHandled` method and call it in two test methods with two different xpaths (unicode and ascii-safe).

For `scrapy.utils.python.unicode_to_str` see https://github.com/scrapy/scrapy/issues/778.

**Comment by Digenis** [comment]
Sure, I'll get to it.
Thanks for pointing out the other issue.

**Comment by kmike** [comment]
Thanks!

**Comment by Digenis** [comment]
Done, (here is a slightly messed diff).
The code duplication tempts me a bit to make a separate method
but these are tests right?, meant to be kept flat... I think.

**Comment by kmike** [comment]
You can move the common logic to a method like `assertInvalidXpathHandled` to avoid duplication. Custom asserts are in line with xUnit philosophy, and a single assert per test is what Java gurus suggest :)

**Comment by kmike** [comment]
I think it is fine to have `exc_msg = str(e) if six.PY3 else str(e).decode("unicode_escape")` line in this assert

File: tests/test_selector.py [hunk 1]
```diff
@@ -213,7 +214,23 @@ def test_invalid_xpath(self):
         try:
             x.xpath(xpath)
         except ValueError as e:
-            assert xpath in str(e), "Exception message does not contain invalid xpath"
+            assert xpath in str(e),\
+                "Exception message does not contain invalid xpath"
```
Comment (kmike) [inline tests/test_selector.py]: self.assertIn will show both values in the error message. I think it is more useful than this custom exception message.

**Comment by Digenis** [comment]
I chose something even shorter

**Comment by kmike** [comment]
:+1:

**Comment by Digenis** [comment]
Is this getting merged all the way to 0.22?
As a bugfix, I think it ought to go as far as all supported versions.

What about @umrashrf's work on #906?

**Comment by kmike** [comment]
@Digenis I don't think we're still supporting 0.22.x. A backport to 0.24.x is more likely, but I'm not sure there will be another 0.24.x release. @dangra @pablohoffman - are there plans to do one more 0.24.x release?

**Comment by dangra** [comment]
Only backport bugfixes to current stable 0.24, it has been our policy so far. +1 for 0.24.6

**Comment by umrashrf** [comment]
> What about @umrashrf's work on #906?

@Digenis What do you mean? Please don't let my PR get in your way if it is! It's a small change so I can cherry-pick it.

**Comment by Digenis** [comment]
No, it doesn't get on my way at all, nor yours.
I thought you are probably keeping track of all recent developments on the selectors, so I just tagged you.
(Quite silly in retrospect since you can find them more easily in git)

**Comment by umrashrf** [comment]
yeah reference is helpful thanks @Digenis

**Comment by Digenis** [comment]
I think I have nothing more left to do on this PR unless you expect me to rebase.

**Comment by dangra** [comment]
please rebase and +1 to merge

**Comment by dangra** [comment]
the change is small enough to squash all commits into one.

**Comment by Digenis** [comment]
re-rebased and squashed

**Comment by kmike** [comment]
Thanks @Digenis!
